### ISSUE TITLE
[Bug]: assigning to `moduleSideEffects` returned from `this.getModuleInfo` has no effect

### ISSUE THREAD

**Issue by sapphi-red** [rolldown/rolldown#2781]
### Reproduction link or steps

1. Open https://stackblitz.com/edit/github-xv4yss-dqmypd?file=rolldown.config.mjs
2. `npm run build` runs automatically
3. Output includes `console.log("no sideeffects");`
4. Run `npm run build:up` to use rollup
5. Output does not include `console.log("no sideeffects");`

### What is expected?

Output to not include `console.log("no sideeffects");`

### What is actually happening?

Output includes `console.log("no sideeffects");`

### System Info

```Shell
Stackblitz

  System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.20.3 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.15.6 - /usr/local/bin/pnpm
  npmPackages:
    rolldown: https://pkg.pr.new/rolldown@9eb7103 => 0.14.0
```

### Any additional comments?

related to #2752

I guess this is hard to fix directly as well.

**Comment by IWANABETHATGUY** [comment]
Yeah, need to proxy the whole `ModuleInfo` and reflect the modification to rust side.

**Comment by hyf0** [comment]
This is important and hard topic of how we share and sync data between rust and js with both ensure the safety and performance.

**Comment by shulaoda** [comment]
Use a `proxy` to update all modifiable data within hooks, and batch sync pending updates to the `rust` side after the relevant hook completes execution if necessary?
